amundsen_application/static/js/ducks/search/filters/reducer.ts

```diff
@@ -0,0 +1,119 @@
+import { FilterInput, ResourceType } from 'interfaces';
+
+export enum UpdateSearchFilter {
+  ADD_MULTI_SELECT = 'amundsen/search/filter/ADD_MULTI_SELECT',
+  REMOVE_MULTI_SELECT = 'amundsen/search/filter/REMOVE_MULTI_SELECT',
+  UPDATE_SINGLE =  'amundsen/search/filter/UPDATE_SINGLE',
+}
+export interface UpdateSearchFilterAction {
+  payload: FilterInput;
+  type: UpdateSearchFilter;
+};
+
+/* ACTIONS */
+export function updateSingleOption(input: FilterInput) {
+  const { category, value } = input;
+  return {
+    payload: {
+      category,
+      value
+    },
+    type: UpdateSearchFilter.UPDATE_SINGLE,
+  };
+};
+export function addMultiSelectOption(input: FilterInput) {
+  const { category, value } = input;
+  return {
+    payload: {
+      category,
+      value
+    },
+    type: UpdateSearchFilter.ADD_MULTI_SELECT,
+  };
+};
+export function removeMultiSelectOption(input: FilterInput) {
+  const { category, value } = input;
+  return {
+    payload: {
+      category,
+      value
+    },
+    type: UpdateSearchFilter.REMOVE_MULTI_SELECT,
+  };
+};
+
+/* REDUCER */
+export type FilterCategories = { [id:string]: string };
+export interface FilterReducerState {
+  [ResourceType.table]: TableFilterReducerState;
+};
+export interface TableFilterReducerState {
+  column: string;
+  database: FilterCategories;
+  schema: string;
+  table: string;
+  tag: string;
+};
+
+export const initialTableFilterState = {
+  column: '',
+  database: {},
+  schema: '',
+  table: '',
+  tag: '',
+};
+
+export const initialFilterState: FilterReducerState = {
+  [ResourceType.table]: initialTableFilterState,
+};
+
+export default function reducer(state: FilterReducerState = initialFilterState, action, resourceType: ResourceType): FilterReducerState {
+  const { category, value  } = action.payload;
```

Comment:
I don't have any ideas at the moment, but we should try find a way to simplify this function.

Reply:
I believe simplification will come with some iterations of this feature. The simplified state suggested above might help.

Reply:
I have a proposal that might simplify things: 
- Each filter component (checkbox, text input) has some attributes such as `name`, `value`, and `checked`. We could leverage each of these form input attributes
- On form submission, the SearchFilters component can iterate over the form elements and generate a new filters json object with some simple rules. 
  - Text Input: If the value is non-empty, we can add a filter. 
  ```
  filters = {};
  ...
  if (textInput.value !== null) 
    filters[textInput.name] = textInput.value
  ```
  - Checkbox: If checked, add the `value` to the `name` category. (Create a new array, or append if it exists). 
  ``` 
  filters = {};
  ...
  if (checkbox.checked) {
    if (filters[checkbox.name] !== undefined) 
      filters[checkbox.name] = [checkbox.value];
    else
      filters[checkbox.name].push(checkbox.value);
  } 
  ```

Reply:
I see. I'm going to explore this one in my other branch `ttannis_db-filter-2`. I had went ahead to do the `SearchFilter` cleanup that involved making shared `FilterSection` component that renders the section header and either `CheckboxFilter` or `InputFilter` as a child and those changes affect how I would try to make this work.

It sounds like I'd have to add support to the `SearchFilter` to get the piece of json for each `FilterSection` and that requires having a `ref` to each `FilterSection` to call the relevant method. I hope that doesn't violate any best practices but I think I already see how much this would clean up the reducer. 

I will explore in `ttannis_db-filter-2` and open a pull request again this branch.

Reply:
This comment is now addressed in work from this PR: https://github.com/lyft/amundsenfrontendlibrary/pull/361